File: cpp-solve/leetcode/sliding_window/s438.cpp
#include "bits/stdc++.h"

using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        map<char, int> window;
        map<char, int> need;
        for (char c:p) need[c]++;
        int valid = 0, l = 0;
        vector<int> ans;
        for (int r = 0; r < s.size(); ++r) {
            char c = s[r];
            if (need.count(c)) {
                window[c]++;
                if (window[c] == need[c]) valid++;
            }
            if (r - l + 1 == p.size()) {
                c = s[l];
                if (valid == need.size()) ans.push_back(l);
                if (need.count(c)) {
                    if (window[c] == need[c]) valid--;
                    window[c]--;
                }
                l++;
            }
        }
        return ans;
    }
};

int main() {
    vector<int> ans = Solution().findAnagrams(
            "cbaebabacd",
            "abc"
    );
    cout << ans.size();
}

File: python/s438.py
import collections
from typing import List


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        need = collections.defaultdict(int)
        window = collections.defaultdict(int)
        need.update(collections.Counter(p))
        res = []
        valid = 0
        l, r = 0, 0
        while r < len(s):
            c = s[r]
            r += 1
            if c in need:
                window[c] += 1
                if window[c] == need[c]:
                    valid += 1
            if r - l == len(p):
                if valid == len(need):
                    res.append(l)
                c = s[l]
                l += 1
                if c in need:
                    if window[c] == need[c]:
                        valid -= 1
                    window[c] -= 1
        return res
